processor/src/trace/virtual_bus/sum_check/domain.rs

```diff
@@ -0,0 +1,65 @@
+use alloc::vec::Vec;
+use vm_core::FieldElement;
+use winter_prover::math::batch_inversion;
+
+/// Implements barycentric evaluation where the interpolation domain is `0, 1, ..., max_degree`
+/// where `max_degree` is maximal polynomial supported degree.
+pub struct EvaluationDomain<E>
+where
+    E: FieldElement,
+{
+    interpolation_points: Vec<E>,
+    barycentric_weights: Vec<E>,
+}
+
+impl<E> EvaluationDomain<E>
+where
+    E: FieldElement,
+{
+    pub fn new(max_degree: usize) -> Self {
+        let interpolation_points: Vec<E> = (0..=max_degree).map(|x| E::from(x as u32)).collect();
+        let barycentric_weights = barycentric_weights(&interpolation_points);
+
+        Self {
+            interpolation_points,
+            barycentric_weights,
+        }
+    }
+
+    pub fn evaluate(&self, evaluations: &[E], r: E) -> E {
+        evaluate_barycentric(&self.interpolation_points, evaluations, r, &self.barycentric_weights)
+    }
+}
+
+/// Computes the barycentric weights for a set of interpolation points.
+pub fn barycentric_weights<E: FieldElement>(points: &[E]) -> Vec<E> {
+    let n = points.len();
+    let tmp = (0..n)
+        .map(|i| (0..n).filter(|&j| j != i).fold(E::ONE, |acc, j| acc * (points[i] - points[j])))
+        .collect::<Vec<_>>();
+    batch_inversion(&tmp)
+}
+
+/// Computes the value of the polynomial with minimal degree interpolating the set `{(x_i, y_i)}`
+/// at the point `r` given pre-computed barycentric weights.
+pub fn evaluate_barycentric<E: FieldElement>(
+    x_i: &[E],
+    y_i: &[E],
```

Comment:
nit: I would rename to `xs` and `ys`, since `x_i` usually stands for "the ith `x`", and it sort of clashes with the iteration variable `i` in `x_i[i]` (and similarly for `y_i`)